### PR TITLE
Add wasm32v1-none target (compiler-team/#791)

### PR DESCRIPTION
**Pull request by graydon**
This is a preliminary implementation of the MCP discussed in [compiler-team#791](https://github.com/rust-lang/compiler-team/issues/791). It's not especially "major" but you know, process! Anyway it adds a new wasm32v1-none target which just pins down a set of wasm features. I think this is close to the consensus that emerged when discussing it on Zulip so I figured I'd sketch to see how hard it is. Turns out not very.

### CODE REVIEW

File: compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
Comment (Urgau) [inline compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:1]: All of this context and explanation is great, it should go as well (or entirely) into a new entry for the target in the `rustc` book:
 - needs to be in the "Tier 2 without host" table, https://github.com/rust-lang/rust/blob/df1b5d3cc2117f1ee96abca25678bc5f5604d450/src/doc/rustc/src/platform-support.md?plain=1#L136
 - with a new file in https://github.com/rust-lang/rust/tree/master/src/doc/rustc/src/platform-support following the [`TEMPLATE`](https://github.com/rust-lang/rust/blob/df1b5d3cc2117f1ee96abca25678bc5f5604d450/src/doc/rustc/src/platform-support/TEMPLATE.md)
 - as well as not forgetting to add the newly created page inside the [`SUMMARY.md`](https://github.com/rust-lang/rust/blob/df1b5d3cc2117f1ee96abca25678bc5f5604d450/src/doc/rustc/src/SUMMARY.md)

*Don't hesitate to take a look at the other wasm targets platform entries[^1][^2].*

[^1]: https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/platform-support/wasm32-wasip1.md
[^2]: https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md

File: compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
Comment (jieyouxu) [inline compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:35]: Suggestion: use the "official" wasm proposal names here, like `reference-types`? Also below.

File: compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
Comment (graydon) [reply compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:1]: Will do, thanks!

Comment (workingjubilee) [reply compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:1]: That will also take care of tidy's whining.

**Review by alexcrichton — COMMENTED** [review]
For tier 2-related bits, some other locations you'll want to modify are:

* [rustup manifest building code](https://github.com/rust-lang/rust/blob/8d94e06ec9758b5c03ea77bb5dab22a1a76bc261/src/tools/build-manifest/src/main.rs#L158-L163)
* [`Dockerfile` used to build other wasm targets](https://github.com/rust-lang/rust/blob/8d94e06ec9758b5c03ea77bb5dab22a1a76bc261/src/ci/docker/host-x86_64/dist-various-2/Dockerfile#L116-L120)

File: compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
Comment (alexcrichton) [reply compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:61]: Oh I reread [the MCP](https://github.com/rust-lang/compiler-team/issues/791) and the answer is "well that's what 1.0 is according to wasm". Given that mind leaving a comment here to that effect?

Comment (graydon) [reply compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:61]: Done.

File: compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
Comment (graydon) [reply compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:35]: Done.

File: compiler/rustc_target/src/spec/targets/wasm32v1_none.rs
Comment (graydon) [reply compiler/rustc_target/src/spec/targets/wasm32v1_none.rs:1]: Done.

**Review by alexcrichton — APPROVED** [review]
Oops sorry I wrote these comments last week but forgot to hit submit...

Otherwise though the MCP has now finished so with this it looks good to me and would be happy to approve.

File: compiler/rustc_target/src/spec/mod.rs
Comment (leighmcculloch) [inline compiler/rustc_target/src/spec/mod.rs:1806]: Will the way to cfg on this specific target be to specify:
```rust
#[cfg(all(target_family = "wasm", target_os = "none"))]
```
or
```rust
#[cfg(all(target_arch = "wasm32", target_os = "none"))]
```

If in the future a `wasm32v2` is released, how will it be distinguished in a `cfg`?

Or will the `target_arch` be `wasm32v1` and in the future `wasm32v2`?

Comment (CryZe) [reply compiler/rustc_target/src/spec/mod.rs:1806]: You can just differentiate them through the target features.

Comment (leighmcculloch) [reply compiler/rustc_target/src/spec/mod.rs:1806]: Could you provide an example?

At the moment differentiating using `target_features` seems unreliable, because it's common for target features to become released but their cfg to be still marked as unstable and is therefore not usable in stable rust, as has been the case with:
- https://github.com/rust-lang/rust/pull/131080

Comment (alexcrichton) [reply compiler/rustc_target/src/spec/mod.rs:1806]: FWIW the documentation in this PR [has an answer to your original question](https://github.com/rust-lang/rust/pull/131487/files#diff-cb13cc0f59e5c9559f00dfaa9afec2b266a09ff54a803ab200a6c1162c3b9c0bR33):

> It's recommended to conditionally compile code for this target with:
>
> ```text
> #[cfg(all(target_family = "wasm", target_os = "none"))]
> ```

Otherwise I would also agree with @CryZe. All other native platforms have a constant `target_arch` string with a variable `target_feature` set corresponding to levels of support for various features.

The situation you're referring to where `multivalue` and `reference-types` aren't stable right now isn't an issue for this target. This'll be a new target when it lands and users should also be able to see `multivalue` and `reference-types` when they get to use it since that PR should land relatively soon as well.

Stabilization of wasm features has lagged behind for awhile and this is basically good motivation to ensure we keep up in the future. For example another option is blocking a hypothetical `wasm32v2-none` target until all of the features it enables are landed and stable in Rust. (or, for example, blocking this PR on #131080 but to be clear I'm not advocating we do that)

Comment (leighmcculloch) [reply compiler/rustc_target/src/spec/mod.rs:1806]: Will the `target_env` be set to `v1` or `v2`?

Comment (workingjubilee) [reply compiler/rustc_target/src/spec/mod.rs:1806]: I do not expect us to update `env` based on the architecture version, no.

Comment (workingjubilee) [reply compiler/rustc_target/src/spec/mod.rs:1806]: As Alex said: The incentive structure that encourages people to push for stabilizing a target feature name that is well-defined, so that we can e.g. evaluate it for its ABI impacts, is intentional.

Comment (graydon) [reply compiler/rustc_target/src/spec/mod.rs:1806]: I think what @leighmcculloch is looking for is a way to make a cfg check that specifically identifies the condition of "being limited to the features in `wasm32v1-none`". Like he wants to not detect any specific new features, but detect the _absence_ of _all_ new features, including features not-yet-invented. Because that's the only case we want to compile for. We want to emit an error if the user tries to build for a target that enables any further wasm features. Which -- even if the current set of feature names all stabilize -- he can't write code to detect because they don't exist yet.

If there were -- speaking hypothetically -- some way to say `cfg(target_triple="wasm32v1-none")` or even `cfg(target_subarch="wasm32v1")` that would do the trick. Or anything else that pins to this subarch version _and not future versions_. I'm not sure if there's a good way to make that happen with the tools currently available though.

File: src/doc/rustc/src/platform-support/wasm32v1-none.md
Comment (graydon) [reply src/doc/rustc/src/platform-support/wasm32v1-none.md:100]: I'll add a note, but I think the build-std part needs to stay if you're targeting MVP and not MVP+mutable-globals as the wasm32v1 target supplies. Or if you need std.

File: compiler/rustc_target/src/spec/mod.rs
Comment (workingjubilee) [reply compiler/rustc_target/src/spec/mod.rs:1806]: I believe there is nonesuch in ways to conveniently do that, yes.

The answer, however, is `cfg(target_feature = "nontrapping-fptoint")`.

The reason that will work is that feature was specifically added **for Rust**, *and* it is part of the v2 definition, so "post v1" Rust targets simply will not ever exclude it.

Comment (graydon) [reply compiler/rustc_target/src/spec/mod.rs:1806]: Mhm. Thanks. I'm sure we'll figure out some not-so-bad guardrails that handle the majority of cases.

**Comment by danielhjacobs** [comment]
We're thinking of switching to this target in our code-base once it reaches stable, but another developer had a question.

> Is webassembly Mutex as stub, or a valid implementation that happens to do nothing?
> not having `std` would make it unusable as-is
> unless I misunderstood it

**Comment by danielhjacobs** [comment]
In testing, I got these errors:

```
error[E0463]: can't find crate for `core`
  |
  = note: the `wasm32-unknown-unknown` target may not be installed
  = help: consider downloading the target with `rustup target add wasm32-unknown-unknown`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

For more information about this error, try `rustc --explain E0463`.
error: could not compile `cfg-if` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error[E0463]: can't find crate for `std`
  |
  = note: the `wasm32-unknown-unknown` target may not be installed
  = help: consider downloading the target with `rustup target add wasm32-unknown-unknown`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`
```

**Comment by alexcrichton** [comment]
@danielhjacobs there's no `std`, so no `Mutex` or `Path` or `Duration`. There's no means by which to re-enable `std` for this target unless you're using `-Zbuild-std`, at which point you can also use `wasm32-unknown-unknown`.

For the build error, I think that's still something using `--target wasm32-unknown-unknown` which needs to switch to `wasm32v1-none`, but if you need `std` to work then this target won't work for you

**Comment by danielhjacobs** [comment]
Hi Alex. That's unfortunate, we were hoping to keep Ruffle working on Pale Moon and on old versions of Safari before the reference types proposal was supported. We can't do that without pinning to Rust 1.81.0 or using nightly Rust, which we don't want to do. https://github.com/rustwasm/wasm-bindgen/pull/4213 will cause wasm-bindgen to at least generate JS glue code that doesn't use those features, but the statement that "For the reference-types feature the encoding of immediates in the call_indirect, a commonly used instruction by the WebAssembly backend, has changed. Validators and parsers which don't understand the reference-types proposal will no longer accept modules produced by LLVM due to this change in encoding of immediates." makes me wary Ruffle will work on those browsers even after that wasm-bindgen PR is out in a new release.

**Comment by danielhjacobs** [comment]
I assume there's no chance to get a target like this but which builds `std`?

**Comment by leighmcculloch** [comment]
> error[E0463]: can't find crate for `core`

Does this mean core is also unavailable on this target?

**Comment by danielhjacobs** [comment]
> Does this mean core is also unavailable on this target?

No, that one was probably my mistake. After installing this target, I failed to update https://github.com/ruffle-rs/ruffle/blob/e81bbe858f07fe8c52e1e30f660eeaea440708c5/web/packages/core/tools/build_wasm.ts. The latest attachment of error messages are the errors I got after changing that file to use this target instead.

**Comment by workingjubilee** [comment]
> I assume there's no chance to get a target like this but which builds `std`?
> 

Probably Not. We do not really want to entertain another target that has a standard library as busted as wasm32-unknown-unknown's is.

> but the statement that "For the reference-types feature the encoding of immediates in the call_indirect, a commonly used instruction by the WebAssembly backend, has changed. Validators and parsers which don't understand the reference-types proposal will no longer accept modules produced by LLVM due to this change in encoding of immediates." makes me wary Ruffle will work on those browsers even after that wasm-bindgen PR is out in a new release.

Using this target, plus the relevant wasm tooling with that "don't enable `externref` automatically" PR merged, should generate correct code for your use-case that is recognized by Pale Moon and old Safari.

The errors in the log you showed are from using crates with features enabled that cause them to not support `no_std`. As the `once_cell` crate supports using it on `no_std`, and I did not see any direct dependencies on it in your tree's Cargo.tomls, it is the result of other crates you depend on.

**Comment by workingjubilee** [comment]
I opened this issue: https://github.com/ruffle-rs/ruffle/issues/18412

I will continue the discussion there.

**Pull request merged**
